mlir/lib/Analysis/Presburger/Simplex.cpp

```diff
@@ -0,0 +1,929 @@
+    result.unknownFromIndex(result.rowVar.back()).pos =
+        result.rowVar.size() - 1;
+    result.nRow++;
+  };
+
+  // Also fixes the corresponding entry in rowVar and var/con.
+  auto appendRowFromB = [&](unsigned row) {
+    result.tableau(result.nRow, 0) = A.tableau(row, 0);
+    result.tableau(result.nRow, 1) = A.tableau(row, 1);
+
+    unsigned offset = A.nCol - 2;
+    for (unsigned col = 2; col < B.nCol; col++)
+      result.tableau(result.nRow, offset + col) = B.tableau(row, col);
+    result.rowVar.push_back(indexFromBIndex(B.rowVar[row]));
+    result.unknownFromIndex(result.rowVar.back()).pos =
+        result.rowVar.size() - 1;
+    result.nRow++;
+  };
+
+  for (unsigned row = 0; row < A.nRow; row++)
+    appendRowFromA(row);
+  for (unsigned row = 0; row < B.nRow; row++)
+    appendRowFromB(row);
+
+  return result;
+}
+
+llvm::Optional<std::vector<int64_t>> Simplex::getSamplePointIfIntegral() const {
+  if (empty)
+    return {};
+
+  std::vector<int64_t> sample;
+  for (const Unknown &u : var) {
+    if (!u.ownsRow)
+      sample.push_back(0);
+    else {
+      if (tableau(u.pos, 1) % tableau(u.pos, 0) != 0)
+        return {};
+      sample.push_back(tableau(u.pos, 1) / tableau(u.pos, 0));
+    }
+  }
+  return sample;
+}
+
+// Given a simplex for a polytope, construct a new simplex whose variables are
+// identified with a pair of points (x, y) in the original polytope. Supports
+// some operations needed for general basis reduction. In what follows, <x, y>
+// denotes the dot product of the vectors x and y.
+class GBRSimplex {
+public:
+  GBRSimplex(const Simplex &originalSimplex)
+      : simplex(Simplex::makeProduct(originalSimplex, originalSimplex)),
+        simplexConstraintOffset(simplex.numberConstraints()) {}
+
+  // Add an equality <dir, x - y> = 0.
+  void addEqualityForDirection(const std::vector<int64_t> &dir) {
+    snapshotStack.push_back(simplex.getSnapshot());
+    simplex.addEquality(getCoeffsForDirection(dir));
+  }
+
+  // Compute max(<dir, x - y>) and save the dual variables for only the
+  // direction constraints to `dual`.
+  Fraction<int64_t> computeWidthAndDuals(const std::vector<int64_t> &dir,
+                                         std::vector<int64_t> &dual,
+                                         int64_t &dualDenom) {
+    unsigned snap = simplex.getSnapshot();
+    unsigned conIndex = simplex.addRow(getCoeffsForDirection(dir));
+    unsigned row = simplex.con[conIndex].pos;
+    auto width = simplex.computeRowOptimum(Simplex::Direction::UP, row);
+    assert(width && "Width should not be unbounded!");
+    dualDenom = simplex.tableau(row, 0);
+    dual.clear();
+    // The increment is i += 2 because equalities are added as two inequalities,
+    // one positive and one negative. We only want to process the positive ones.
+    for (unsigned i = simplexConstraintOffset; i < conIndex; i += 2) {
+      if (simplex.con[i].ownsRow)
+        dual.push_back(0);
+      else {
```

Comment:
Here I would probably drop the brackets, but this is a discussion of taste I guess

Reply:
Hm, unless this is generally followed in MLIR I prefer to keep the braces.